In [11]:
import pandas as pd

df = pd.read_csv('rubik_solves.csv')

df = df.replace(['n/a', 'None', 'N/A', 'null'], pd.NA)

df_cleaned = df.dropna(subset=['solver', 'pll_time', 'pll'])

print(f"Удалено строк: {len(df) - len(df_cleaned)}")
print(f"Осталось строк: {len(df_cleaned)}")
df_cleaned.head()

Удалено строк: 289
Осталось строк: 711


,solve_id,solver,solve_time,pll,pll_time,Split,STM,STPS,ETM,ETPS
4,12922,Radosław Marcinek,5.18,R' U' R U D' R2 U R' U R U' R U' R2' U D,1.31,25.3%,16,12.21,16,12.21
5,12921,Dylan Miller,4.40,U U' R2 U R' U R' U' R U' R2 D U' R' U R D' U2',1.37,31.1%,18,13.14,18,13.14
6,12920,Dylan Miller,4.60,U' R' U2 R' D' R U' R' D R U R U' R' U' R,1.03,22.4%,16,15.53,16,15.53
7,12919,Dylan Miller,5.26,U' R2 U R' U R' U' R U' R2 U' D R' U R D' U,1.53,29.1%,17,11.11,17,11.11
8,12918,Dylan Miller,6.72,U U' R' U' R U D' R2 U R' U R U' R U' R2' U' D U',1.95,29.0%,19,9.74,19,9.74


Осталось 711 объектов. При необходимости можно будет напарсить больше данных

In [12]:
df=df_cleaned

In [13]:
for move in "ULFRBD":
  df[f"{move}"]=df["pll"].apply(lambda x: x.count(f"{move}")-x.count(f"{move}2"))
  df[f"{move}2"]=df["pll"].apply(lambda x: x.count(f"{move}2"))
df.head()

,solve_id,solver,solve_time,pll,pll_time,Split,STM,STPS,ETM,ETPS,...,L,L2,F,F2,R,R2,B,B2,D,D2
4,12922,Radosław Marcinek,5.18,R' U' R U D' R2 U R' U R U' R U' R2' U D,1.31,25.3%,16,12.21,16,12.21,...,0,0,0,0,5,2,0,0,2,0
5,12921,Dylan Miller,4.40,U U' R2 U R' U R' U' R U' R2 D U' R' U R D' U2',1.37,31.1%,18,13.14,18,13.14,...,0,0,0,0,5,2,0,0,2,0
6,12920,Dylan Miller,4.60,U' R' U2 R' D' R U' R' D R U R U' R' U' R,1.03,22.4%,16,15.53,16,15.53,...,0,0,0,0,8,0,0,0,2,0
7,12919,Dylan Miller,5.26,U' R2 U R' U R' U' R U' R2 U' D R' U R D' U,1.53,29.1%,17,11.11,17,11.11,...,0,0,0,0,5,2,0,0,2,0
8,12918,Dylan Miller,6.72,U U' R' U' R U D' R2 U R' U R U' R U' R2' U' D U',1.95,29.0%,19,9.74,19,9.74,...,0,0,0,0,5,2,0,0,2,0


Создаю дополнительные фичи. Какие ходы и комбинации ходов присутствуют в решении.

In [14]:
df["simul"]=df["pll"].apply(lambda x: x.count("U D")+x.count("U' D")+x.count("U2 D")+x.count("D U")+x.count("D' U")+x.count("D2 U"))

In [15]:
df["ghost_move"]=df["pll"].apply(lambda x: x.count("R' U R'") + x.count("L U' L"))
df.head()

,solve_id,solver,solve_time,pll,pll_time,Split,STM,STPS,ETM,ETPS,...,F,F2,R,R2,B,B2,D,D2,simul,ghost_move
4,12922,Radosław Marcinek,5.18,R' U' R U D' R2 U R' U R U' R U' R2' U D,1.31,25.3%,16,12.21,16,12.21,...,0,0,5,2,0,0,2,0,2,0
5,12921,Dylan Miller,4.40,U U' R2 U R' U R' U' R U' R2 D U' R' U R D' U2',1.37,31.1%,18,13.14,18,13.14,...,0,0,5,2,0,0,2,0,2,1
6,12920,Dylan Miller,4.60,U' R' U2 R' D' R U' R' D R U R U' R' U' R,1.03,22.4%,16,15.53,16,15.53,...,0,0,8,0,0,0,2,0,0,0
7,12919,Dylan Miller,5.26,U' R2 U R' U R' U' R U' R2 U' D R' U R D' U,1.53,29.1%,17,11.11,17,11.11,...,0,0,5,2,0,0,2,0,2,1
8,12918,Dylan Miller,6.72,U U' R' U' R U D' R2 U R' U R U' R U' R2' U' D U',1.95,29.0%,19,9.74,19,9.74,...,0,0,5,2,0,0,2,0,3,0


Получили выборку 711 объектов и 21 фичу(3 столбца не участвуют)
